internal/contracts/manager.go

```diff
@@ -811,7 +840,118 @@ func (cm *contractManager) checkContractListenerExists(ctx context.Context, list
+				return err
+			}
+		} else {
+			filter.Interface = nil
+		}
+
+		if err := cm.validateFFIEvent(ctx, &filter.Event.FFIEventDefinition); err != nil {
+			return err
+		}
+
+		filter.Signature = cm.blockchain.GenerateEventSignature(ctx, &filter.Event.FFIEventDefinition)
+
+		if filter.Location != nil {
+			if filter.Location, err = cm.blockchain.NormalizeContractLocation(ctx, blockchain.NormalizeListener, filter.Location); err != nil {
+				return err
+			}
+		}
+
+		// Have parsed a filter with the same signature?
+		// In the case of location being different
+		if duplicateMapChecker[filter.Signature] != "" {
+			// If this duplicate filter is looking at all locations it's a superset of the previous one
+			// or the previous filter was also looking at all locations then we are trying to add a subset
+			if filter.Location == nil || duplicateMapChecker[filter.Signature] == "*" {
+				return i18n.NewError(ctx, coremsgs.MsgDuplicateContractListenerFilterLocation)
+			}
+
+			location, err := cm.blockchain.StringifyContractLocation(ctx, filter.Location)
+			if err != nil {
+				return err
+			}
+
+			// Check if same location
+			if strings.Compare(duplicateMapChecker[filter.Signature], location) == 0 {
+				return i18n.NewError(ctx, coremsgs.MsgDuplicateContractListenerFilterLocation, location)
+			}
+
+			// In the case of Fabric for example comparing listening to channel
+			// comparing to channel + chaincode
+			if strings.Contains(duplicateMapChecker[filter.Signature], location) || strings.Contains(location, duplicateMapChecker[filter.Signature]) {
+				return i18n.NewError(ctx, coremsgs.MsgDuplicateContractListenerFilterLocation)
+			}
+
+		}
+
+		listener.ContractListener.Filters = append(listener.ContractListener.Filters, &core.ListenerFilter{
+			Event:     filter.Event,
+			Location:  filter.Location,
+			Interface: filter.Interface,
+			Signature: filter.Signature,
+		})
+
+		location := "*"
+		if filter.Location != nil {
+			location, err = cm.blockchain.StringifyContractLocation(ctx, filter.Location)
+			if err != nil {
+				return err
+			}
+		}
+		duplicateMapChecker[filter.Signature] = location
+	}
+
+	// To preserve compatibility copy the first element in the
+	// filters arary to the top if not present already
+	if listener.Event == nil && len(listener.Filters) > 0 {
+		listener.Event = listener.Filters[0].Event
+		listener.Location = listener.Filters[0].Location
+		listener.Interface = listener.Filters[0].Interface
+	}
+
+	return nil
+}
+
+func (cm *contractManager) ConstructContractListenerSignature(ctx context.Context, listener *core.ContractListenerInput) (output *core.ContractListenerSignatureOutput, err error) {
+	output = &core.ContractListenerSignatureOutput{}
+
+	err = cm.parseContractListenerFilters(ctx, listener)
+	if err != nil {
+		return nil, err
+	}
+
+	output.Signature, err = cm.generateFilterSignature(ctx, listener.ContractListener.Filters)
```

Comment:
Why is this a separate function to `parseContractListenerFilters`?
It seems like `parseContractListenerFilters` already updates the output to set `Event` and `Location` - so why wouldn't it set `Signature` too?

Reply:
Done